fix: deploy CI frontend validation fails with curl exit 6 due to CloudFront canonical-domain redirect

## Root cause

The **"Validate backend and frontend endpoints"** step in `deploy-lambda.yml` fails with:

```
curl failed (exit 6) checking https://d32n5ua14kb82t.cloudfront.net
```

The failure chain:

1. `getent ahostsv4` correctly resolves the CloudFront domain to an IP (e.g. `18.172.124.34`).
2. `curl --resolve "${frontend_domain}:443:${frontend_ip}" --location https://${frontend_domain}` sends the request.
3. The CloudFront viewer-request function (`cdk/functions/viewer-request.js`) sees `host !== 'app.allotmint.io'` and returns **301 → `https://app.allotmint.io/`**.
4. `--location` tells curl to follow that redirect.
5. curl's c-ares resolver tries to look up `app.allotmint.io` — but `--resolve` only pinned `d32n5ua14kb82t.cloudfront.net:443`, so c-ares is used for the redirect target.
6. c-ares fails to resolve `app.allotmint.io` → **exit code 6**.
7. `--retry-all-errors` retries 3× with 5 s delay (~15 s total), same outcome each time.

The PR #3003 fix (bypassing c-ares via `--resolve` + `getent ahostsv4`) correctly handles the *initial* request, but the redirect to the canonical domain bypasses the pinned IP because `--resolve` only covers the distribution domain, not `app.allotmint.io`.

## Fix

Remove `--location` from the CloudFront curl check. The distribution is behaving correctly — returning 301 to the canonical domain is expected. The check should accept **301** (or any 3xx) as a valid "distribution is up" response, instead of trying to follow the redirect to the canonical domain.

**Files to change:** `.github/workflows/deploy-lambda.yml`

- Remove `--location` from the `cf_status=$(curl ...)` command
- Change the success condition from `== "200"` to also accept `"301"` (the expected redirect from the raw CloudFront domain)

## Logs

Run ID: 69416311599 — step 24 "Validate backend and frontend endpoints"

```
DNS resolved to 18.172.124.34 (attempt 1/12)
Checking frontend → https://d32n5ua14kb82t.cloudfront.net
curl failed (exit 6) checking https://d32n5ua14kb82t.cloudfront.net
Error: Process completed with exit code 1.
```

--- a/.github/workflows/deploy-lambda.yml
+++ b/.github/workflows/deploy-lambda.yml
@@ -307,9 +307,22 @@ jobs:
             echo "DNS for ${frontend_domain} still not resolving after 120 s" >&2
             exit 1
           fi
-          canonical_ip="$(getent ahostsv4 "${canonical_domain}" 2>/dev/null | awk '{print $1; exit}')" || true
+          # app.allotmint.io is a CNAME to the same distribution, so it should resolve
+          # as soon as frontend_domain does — but a short retry guards against the
+          # same transient DNS blips that the loop above handles for frontend_domain.
+          canonical_ip=""
+          for i in $(seq 1 4); do
+            canonical_ip="$(getent ahostsv4 "${canonical_domain}" 2>/dev/null | awk '{print $1; exit}')" || true
+            if [ -n "${canonical_ip}" ]; then
+              break
+            fi
+            if [ "$i" -lt 4 ]; then
+              echo "Attempt $i/4: ${canonical_domain} DNS not yet resolving; retrying in 5 s..."
+              sleep 5
+            fi
+          done
           if [ -z "${canonical_ip}" ]; then
-            echo "DNS for ${canonical_domain} is not resolving" >&2
+            echo "DNS for ${canonical_domain} is not resolving after 4 attempts" >&2
             exit 1
           fi
           echo "DNS resolved ${canonical_domain} to ${canonical_ip}"

--- a/.github/workflows/deploy-lambda.yml
+++ b/.github/workflows/deploy-lambda.yml
@@ -277,13 +277,14 @@ jobs:
           # two backends cannot disagree. See issues #3000, #3002, and #3004.
           # The viewer-request CloudFront function (cdk/functions/viewer-request.js)
           # redirects any request whose Host is not app.allotmint.io to the canonical
-          # host with 301. Resolving the distribution domain (*.cloudfront.net) and then
-          # curling the canonical URL pins c-ares out of the picture entirely: the IP is
-          # known from getent, --resolve maps it for the canonical hostname, and the
-          # viewer-request function passes the request through to S3 rather than
-          # redirecting, so we get a real 200 that validates actual content delivery.
-          # app.allotmint.io is a CNAME to the same CloudFront distribution, so the
-          # IP resolved for the distribution domain is valid for the canonical domain too.
+          # host with 301. To avoid following that redirect (which requires c-ares to
+          # resolve the canonical host and fails in CI), we resolve the canonical domain
+          # independently via getent and curl it directly with --resolve. This validates
+          # actual content delivery (viewer-request passes the request through to S3 →
+          # 200) rather than just distribution reachability (a 301 pre-origin redirect).
+          # The distribution DNS wait loop below guards against a newly created
+          # distribution not yet propagating; app.allotmint.io is a stable CNAME so a
+          # single getent lookup (no retry loop) is sufficient for it.
           canonical_domain="app.allotmint.io"
           echo "Waiting for DNS resolution of ${frontend_domain}..."
           dns_ok=0
@@ -293,7 +294,7 @@ jobs:
             # curl --resolve with IPv6. || true: set -e exits on getent non-zero.
             frontend_ip="$(getent ahostsv4 "$frontend_domain" 2>/dev/null | awk '{print $1; exit}')" || true
             if [ -n "$frontend_ip" ]; then
-              echo "DNS resolved to ${frontend_ip} (attempt $i/12)"
+              echo "DNS resolved ${frontend_domain} to ${frontend_ip} (attempt $i/12)"
               dns_ok=1
               break
             fi
@@ -306,11 +307,17 @@ jobs:
             echo "DNS for ${frontend_domain} still not resolving after 120 s" >&2
             exit 1
           fi
+          canonical_ip="$(getent ahostsv4 "${canonical_domain}" 2>/dev/null | awk '{print $1; exit}')" || true
+          if [ -z "${canonical_ip}" ]; then
+            echo "DNS for ${canonical_domain} is not resolving" >&2
+            exit 1
+          fi
+          echo "DNS resolved ${canonical_domain} to ${canonical_ip}"
           echo "Checking frontend → https://${canonical_domain}"
           curl_exit=0
           cf_status=$(curl --retry 3 --retry-delay 5 --retry-all-errors \
             --max-time 60 --write-out "%{http_code}" --silent --output /dev/null \
-            --resolve "${canonical_domain}:443:${frontend_ip}" \
+            --resolve "${canonical_domain}:443:${canonical_ip}" \
             "https://${canonical_domain}") || curl_exit=$?
           if [ "$curl_exit" -ne 0 ]; then
             echo "curl failed (exit ${curl_exit}) checking https://${canonical_domain}" >&2

--- a/.github/workflows/deploy-lambda.yml
+++ b/.github/workflows/deploy-lambda.yml
@@ -275,16 +275,9 @@ jobs:
           # systemd-resolved stub; curl → c-ares async resolver). Capturing the IP here
           # and passing it to curl via --resolve bypasses curl's resolver entirely so the
           # two backends cannot disagree. See issues #3000, #3002, and #3004.
-          # The viewer-request CloudFront function (cdk/functions/viewer-request.js)
-          # redirects any request whose Host is not app.allotmint.io to the canonical
-          # host with 301. To avoid following that redirect (which requires c-ares to
-          # resolve the canonical host and fails in CI), we resolve the canonical domain
-          # independently via getent and curl it directly with --resolve. This validates
-          # actual content delivery (viewer-request passes the request through to S3 →
-          # 200) rather than just distribution reachability (a 301 pre-origin redirect).
-          # The distribution DNS wait loop below guards against a newly created
-          # distribution not yet propagating; app.allotmint.io is a stable CNAME so a
-          # single getent lookup (no retry loop) is sufficient for it.
+          # viewer-request (cdk/functions/viewer-request.js) returns 301 for the raw
+          # distribution domain; curl the canonical domain directly so the function
+          # passes the request to S3 and returns a real 200. See issue #3004.
           canonical_domain="app.allotmint.io"
           echo "Waiting for DNS resolution of ${frontend_domain}..."
           dns_ok=0
@@ -307,22 +300,22 @@ jobs:
             echo "DNS for ${frontend_domain} still not resolving after 120 s" >&2
             exit 1
           fi
-          # app.allotmint.io is a CNAME to the same distribution, so it should resolve
-          # as soon as frontend_domain does — but a short retry guards against the
-          # same transient DNS blips that the loop above handles for frontend_domain.
+          # app.allotmint.io is a CNAME whose Route53 record CDK updates as part of
+          # StaticSiteStack. Give it the same 120 s budget as the distribution domain
+          # in case CNAME propagation lags on a fresh deploy.
           canonical_ip=""
-          for i in $(seq 1 4); do
+          for i in $(seq 1 12); do
             canonical_ip="$(getent ahostsv4 "${canonical_domain}" 2>/dev/null | awk '{print $1; exit}')" || true
             if [ -n "${canonical_ip}" ]; then
               break
             fi
-            if [ "$i" -lt 4 ]; then
-              echo "Attempt $i/4: ${canonical_domain} DNS not yet resolving; retrying in 5 s..."
-              sleep 5
+            if [ "$i" -lt 12 ]; then
+              echo "Attempt $i/12: ${canonical_domain} DNS not yet resolving; retrying in 10 s..."
+              sleep 10
             fi
           done
           if [ -z "${canonical_ip}" ]; then
-            echo "DNS for ${canonical_domain} is not resolving after 4 attempts" >&2
+            echo "DNS for ${canonical_domain} is not resolving after 120 s" >&2
             exit 1
           fi
           echo "DNS resolved ${canonical_domain} to ${canonical_ip}"

--- a/.github/workflows/deploy-lambda.yml
+++ b/.github/workflows/deploy-lambda.yml
@@ -274,13 +274,17 @@ jobs:
           # getent and curl use different DNS backends on Ubuntu 24.04 (getent → NSS/
           # systemd-resolved stub; curl → c-ares async resolver). Capturing the IP here
           # and passing it to curl via --resolve bypasses curl's resolver entirely so the
-          # two backends cannot disagree. See issues #3000 and #3002.
-          # Do NOT use --location: the viewer-request CloudFront function redirects
-          # requests from the raw *.cloudfront.net domain to the canonical host
-          # (app.allotmint.io) with 301. Following that redirect causes curl to look up
-          # the canonical host via c-ares which cannot resolve it in CI → exit 6.
-          # A 301 from the distribution means CloudFront is up and the function is
-          # running correctly, so we accept both 200 and 301. See issue #3004.
+          # two backends cannot disagree. See issues #3000, #3002, and #3004.
+          # The viewer-request CloudFront function (cdk/functions/viewer-request.js)
+          # redirects any request whose Host is not app.allotmint.io to the canonical
+          # host with 301. Resolving the distribution domain (*.cloudfront.net) and then
+          # curling the canonical URL pins c-ares out of the picture entirely: the IP is
+          # known from getent, --resolve maps it for the canonical hostname, and the
+          # viewer-request function passes the request through to S3 rather than
+          # redirecting, so we get a real 200 that validates actual content delivery.
+          # app.allotmint.io is a CNAME to the same CloudFront distribution, so the
+          # IP resolved for the distribution domain is valid for the canonical domain too.
+          canonical_domain="app.allotmint.io"
           echo "Waiting for DNS resolution of ${frontend_domain}..."
           dns_ok=0
           frontend_ip=""
@@ -302,18 +306,18 @@ jobs:
             echo "DNS for ${frontend_domain} still not resolving after 120 s" >&2
             exit 1
           fi
-          echo "Checking frontend → https://${frontend_domain}"
+          echo "Checking frontend → https://${canonical_domain}"
           curl_exit=0
           cf_status=$(curl --retry 3 --retry-delay 5 --retry-all-errors \
             --max-time 60 --write-out "%{http_code}" --silent --output /dev/null \
-            --resolve "${frontend_domain}:443:${frontend_ip}" \
-            "https://${frontend_domain}") || curl_exit=$?
+            --resolve "${canonical_domain}:443:${frontend_ip}" \
+            "https://${canonical_domain}") || curl_exit=$?
           if [ "$curl_exit" -ne 0 ]; then
-            echo "curl failed (exit ${curl_exit}) checking https://${frontend_domain}" >&2
+            echo "curl failed (exit ${curl_exit}) checking https://${canonical_domain}" >&2
             exit 1
           fi
-          if [ "$cf_status" != "200" ] && [ "$cf_status" != "301" ]; then
-            echo "Expected HTTP 200 or 301 from CloudFront frontend, got $cf_status" >&2
+          if [ "$cf_status" != "200" ]; then
+            echo "Expected HTTP 200 from CloudFront frontend, got $cf_status" >&2
             exit 1
           fi
           echo "Frontend → $cf_status OK"

--- a/.github/workflows/deploy-lambda.yml
+++ b/.github/workflows/deploy-lambda.yml
@@ -275,6 +275,12 @@ jobs:
           # systemd-resolved stub; curl → c-ares async resolver). Capturing the IP here
           # and passing it to curl via --resolve bypasses curl's resolver entirely so the
           # two backends cannot disagree. See issues #3000 and #3002.
+          # Do NOT use --location: the viewer-request CloudFront function redirects
+          # requests from the raw *.cloudfront.net domain to the canonical host
+          # (app.allotmint.io) with 301. Following that redirect causes curl to look up
+          # the canonical host via c-ares which cannot resolve it in CI → exit 6.
+          # A 301 from the distribution means CloudFront is up and the function is
+          # running correctly, so we accept both 200 and 301. See issue #3004.
           echo "Waiting for DNS resolution of ${frontend_domain}..."
           dns_ok=0
           frontend_ip=""
@@ -298,7 +304,7 @@ jobs:
           fi
           echo "Checking frontend → https://${frontend_domain}"
           curl_exit=0
-          cf_status=$(curl --location --retry 3 --retry-delay 5 --retry-all-errors \
+          cf_status=$(curl --retry 3 --retry-delay 5 --retry-all-errors \
             --max-time 60 --write-out "%{http_code}" --silent --output /dev/null \
             --resolve "${frontend_domain}:443:${frontend_ip}" \
             "https://${frontend_domain}") || curl_exit=$?
@@ -306,8 +312,8 @@ jobs:
             echo "curl failed (exit ${curl_exit}) checking https://${frontend_domain}" >&2
             exit 1
           fi
-          if [ "$cf_status" != "200" ]; then
-            echo "Expected HTTP 200 from CloudFront frontend, got $cf_status" >&2
+          if [ "$cf_status" != "200" ] && [ "$cf_status" != "301" ]; then
+            echo "Expected HTTP 200 or 301 from CloudFront frontend, got $cf_status" >&2
             exit 1
           fi
           echo "Frontend → $cf_status OK"

--- a/.github/workflows/deploy-lambda.yml
+++ b/.github/workflows/deploy-lambda.yml
@@ -278,17 +278,19 @@ jobs:
           # viewer-request (cdk/functions/viewer-request.js) returns 301 for the raw
           # distribution domain; curl the canonical domain directly so the function
           # passes the request to S3 and returns a real 200. See issue #3004.
+          # app.allotmint.io CNAMEs to the distribution, so waiting for it to resolve
+          # implicitly gates on distribution DNS propagation too (CNAME → distribution
+          # domain → CloudFront edge). Budget matches the old distribution-domain loop.
           canonical_domain="app.allotmint.io"
-          echo "Waiting for DNS resolution of ${frontend_domain}..."
-          dns_ok=0
-          frontend_ip=""
+          echo "Distribution domain: ${frontend_domain}"
+          echo "Waiting for DNS resolution of ${canonical_domain}..."
+          canonical_ip=""
           for i in $(seq 1 12); do
             # ahostsv4 forces IPv4; avoids bracket-notation requirement for
             # curl --resolve with IPv6. || true: set -e exits on getent non-zero.
-            frontend_ip="$(getent ahostsv4 "$frontend_domain" 2>/dev/null | awk '{print $1; exit}')" || true
-            if [ -n "$frontend_ip" ]; then
-              echo "DNS resolved ${frontend_domain} to ${frontend_ip} (attempt $i/12)"
-              dns_ok=1
+            canonical_ip="$(getent ahostsv4 "${canonical_domain}" 2>/dev/null | awk '{print $1; exit}')" || true
+            if [ -n "${canonical_ip}" ]; then
+              echo "DNS resolved ${canonical_domain} to ${canonical_ip} (attempt $i/12)"
               break
             fi
             if [ "$i" -lt 12 ]; then
@@ -296,29 +298,10 @@ jobs:
               sleep 10
             fi
           done
-          if [ "$dns_ok" -eq 0 ]; then
-            echo "DNS for ${frontend_domain} still not resolving after 120 s" >&2
-            exit 1
-          fi
-          # app.allotmint.io is a CNAME whose Route53 record CDK updates as part of
-          # StaticSiteStack. Give it the same 120 s budget as the distribution domain
-          # in case CNAME propagation lags on a fresh deploy.
-          canonical_ip=""
-          for i in $(seq 1 12); do
-            canonical_ip="$(getent ahostsv4 "${canonical_domain}" 2>/dev/null | awk '{print $1; exit}')" || true
-            if [ -n "${canonical_ip}" ]; then
-              break
-            fi
-            if [ "$i" -lt 12 ]; then
-              echo "Attempt $i/12: ${canonical_domain} DNS not yet resolving; retrying in 10 s..."
-              sleep 10
-            fi
-          done
           if [ -z "${canonical_ip}" ]; then
             echo "DNS for ${canonical_domain} is not resolving after 120 s" >&2
             exit 1
           fi
-          echo "DNS resolved ${canonical_domain} to ${canonical_ip}"
           echo "Checking frontend → https://${canonical_domain}"
           curl_exit=0
           cf_status=$(curl --retry 3 --retry-delay 5 --retry-all-errors \

